File: sdk/credentials.go
package sdk

import (
	"fmt"
	"os"
	"path/filepath"

	"gopkg.in/yaml.v2"
)

func loadConfig() Config {
	config := Config{}
	homeDir, err := os.UserHomeDir()
	if err == nil {
		configPath := filepath.Join(homeDir, ".beamlit", "config.yaml")
		if data, err := os.ReadFile(configPath); err == nil {
			if err := yaml.Unmarshal(data, &config); err != nil {
				// Invalid YAML, use empty credentials
				config = Config{}
			}
		}
	}
	return config
}

func saveConfig(config Config) {
	yamlData, err := yaml.Marshal(config)
	if err != nil {
		panic(err)
	}
	homeDir, err := os.UserHomeDir()
	if err != nil {
		fmt.Printf("Error getting home directory: %v\n", err)
		os.Exit(1)
	}
	configDir := filepath.Join(homeDir, ".beamlit")
	configFile := filepath.Join(configDir, "config.yaml")
	if err := os.WriteFile(configFile, yamlData, 0600); err != nil {
		fmt.Printf("Error writing credentials file: %v\n", err)
		os.Exit(1)
	}
}

func ListWorkspaces() []string {
	config := loadConfig()
	workspaces := []string{}
	for _, workspace := range config.Workspaces {
		workspaces = append(workspaces, workspace.Name)
	}
	return workspaces
}

func CurrentContext() ContextConfig {
	config := loadConfig()
	return config.Context
}

func SetCurrentWorkspace(workspaceName string, environment string) {
	config := loadConfig()
	// Check if workspace exists
	found := false
	for _, workspace := range config.Workspaces {
		if workspace.Name == workspaceName {
			found = true
			break
		}
	}
	if !found {
		fmt.Printf("Workspace %s not found\n", workspaceName)
		os.Exit(1)
	}
	config.Context.Workspace = workspaceName
	config.Context.Environment = environment
	saveConfig(config)
}

func LoadCredentials(workspaceName string) Credentials {
	config := loadConfig()
	for _, workspace := range config.Workspaces {
		if workspace.Name == workspaceName {
			return workspace.Credentials
		}
	}
	return Credentials{}
}

func createHomeDirIfMissing() {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		fmt.Printf("Error getting home directory: %v\n", err)
		return
	}

	credentialsDir := filepath.Join(homeDir, ".beamlit")
	credentialsFile := filepath.Join(credentialsDir, "credentials.json")

	// Check if credentials file exists
	if _, err := os.Stat(credentialsFile); err == nil {
		fmt.Println("You are already logged in. Enter a new API key to overwrite it.")
	} else {
		if err := os.MkdirAll(credentialsDir, 0700); err != nil {
			fmt.Printf("Error creating credentials directory: %v\n", err)
			return
		}
	}
}

func SaveCredentials(workspaceName string, credentials Credentials) {
	createHomeDirIfMissing()
	if credentials.AccessToken == "" && credentials.APIKey == "" && credentials.ClientCredentials == "" {
		fmt.Println("No credentials to save, error")
		return
	}
	config := loadConfig()
	found := false
	for i, workspace := range config.Workspaces {
		if workspace.Name == workspaceName {
			config.Workspaces[i].Credentials = credentials
			found = true
			break
		}
	}
	if !found {
		config.Workspaces = append(config.Workspaces, WorkspaceConfig{Name: workspaceName, Credentials: credentials})
	}
	saveConfig(config)
}

func ClearCredentials(workspaceName string) {
	config := loadConfig()
	for i, workspace := range config.Workspaces {
		if workspace.Name == workspaceName {
			config.Workspaces = append(config.Workspaces[:i], config.Workspaces[i+1:]...)
			break
		}
	}
	if config.Context.Workspace == workspaceName {
		config.Context.Workspace = ""
	}
	saveConfig(config)
}

File: sdk-ts/src/authentication/credentials.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "fs";
import * as yaml from "js-yaml";
import { homedir } from "os";
import { join } from "path";
import { Config, ContextConfig, Credentials, WorkspaceConfig } from "./types";

function loadConfig(): Config {
  const config: Config = {
    workspaces: [],
    context: {
      workspace: "",
      environment: "",
    },
  };

  const homeDir = homedir();
  if (homeDir) {
    const configPath = join(homeDir, ".beamlit", "config.yaml");
    if (existsSync(configPath)) {
      try {
        const data = yaml.load(readFileSync(configPath, "utf8")) as any;
        if (data) {
          const workspaces: WorkspaceConfig[] = [];
          for (const ws of data.workspaces || []) {
            const creds: Credentials = {
              apiKey: ws.credentials?.apiKey || "",
              access_token: ws.credentials?.access_token || "",
              refresh_token: ws.credentials?.refresh_token || "",
              expires_in: ws.credentials?.expires_in || 0,
              device_code: ws.credentials?.device_code || "",
              client_credentials: ws.credentials?.client_credentials || "",
            };
            workspaces.push({ name: ws.name, credentials: creds });
          }
          config.workspaces = workspaces;
          if (data.context) {
            config.context = data.context;
          }
        }
      } catch (e) {
        // Invalid YAML, use empty config
      }
    }
  }
  return config;
}

function saveConfig(config: Config): void {
  const homeDir = homedir();
  if (!homeDir) {
    throw new Error("Could not determine home directory");
  }

  const configDir = join(homeDir, ".beamlit");
  const configFile = join(configDir, "config.yaml");

  mkdirSync(configDir, { recursive: true, mode: 0o700 });
  writeFileSync(
    configFile,
    yaml.dump({
      workspaces: config.workspaces.map((ws) => ({
        name: ws.name,
        credentials: {
          apiKey: ws.credentials.apiKey,
          access_token: ws.credentials.access_token,
          refresh_token: ws.credentials.refresh_token,
          expires_in: ws.credentials.expires_in,
          device_code: ws.credentials.device_code,
        },
      })),
      context: config.context,
    })
  );
}

export function listWorkspaces(): string[] {
  const config = loadConfig();
  return config.workspaces.map((workspace) => workspace.name);
}

export function currentContext(): ContextConfig {
  const config = loadConfig();
  return config.context;
}

export function setCurrentWorkspace(
  workspaceName: string,
  environment: string
): void {
  const config = loadConfig();
  config.context.workspace = workspaceName;
  config.context.environment = environment;
  saveConfig(config);
}

export function loadCredentials(workspaceName: string): Credentials {
  const config = loadConfig();
  const workspace = config.workspaces.find((ws) => ws.name === workspaceName);
  if (workspace) {
    return workspace.credentials;
  }
  return {
    apiKey: "",
    access_token: "",
    refresh_token: "",
    expires_in: 0,
    device_code: "",
    client_credentials: "",
  };
}

export function loadCredentialsFromSettings(settings: any): Credentials {
  return {
    apiKey: settings.authentication?.apiKey || "",
    access_token: "",
    refresh_token: "",
    expires_in: 0,
    device_code: "",
    client_credentials: "",
  };
}

export function createHomeDirIfMissing(): void {
  const homeDir = homedir();
  if (!homeDir) {
    console.error("Error getting home directory");
    return;
  }

  const credentialsDir = join(homeDir, ".beamlit");
  const credentialsFile = join(credentialsDir, "credentials.json");

  if (existsSync(credentialsFile)) {
    console.warn(
      "You are already logged in. Enter a new API key to overwrite it."
    );
  } else {
    try {
      mkdirSync(credentialsDir, { recursive: true, mode: 0o700 });
    } catch (e) {
      console.error(`Error creating credentials directory: ${e}`);
    }
  }
}

export function saveCredentials(
  workspaceName: string,
  credentials: Credentials
): void {
  createHomeDirIfMissing();
  if (!credentials.access_token && !credentials.apiKey) {
    console.info("No credentials to save, error");
    return;
  }

  const config = loadConfig();
  let found = false;

  for (let i = 0; i < config.workspaces.length; i++) {
    if (config.workspaces[i].name === workspaceName) {
      config.workspaces[i].credentials = credentials;
      found = true;
      break;
    }
  }

  if (!found) {
    config.workspaces.push({ name: workspaceName, credentials });
  }

  saveConfig(config);
}

export function clearCredentials(workspaceName: string): void {
  const config = loadConfig();
  config.workspaces = config.workspaces.filter(
    (ws) => ws.name !== workspaceName
  );

  if (config.context.workspace === workspaceName) {
    config.context.workspace = "";
  }

  saveConfig(config);
}
